server/app/services/applicant/ApplicantServiceImpl.java

```diff
@@ -358,6 +367,138 @@ public ImmutableList<Application> getAllApplications() {
     return applicationRepository.getAllApplications();
   }
 
+  @Override
+  public CompletionStage<RelevantPrograms> relevantProgramsForApplicant(long applicantId) {
+    // Note: The Program model associated with the application is eagerly loaded.
+    CompletableFuture<ImmutableSet<Application>> applicationsFuture =
+        applicationRepository
+            .getApplicationsForApplicant(
+                applicantId, ImmutableSet.of(LifecycleStage.DRAFT, LifecycleStage.ACTIVE))
+            .toCompletableFuture();
+    ImmutableList<ProgramDefinition> activePrograms =
+        versionRepository.getActiveVersion().getPrograms().stream()
+            .map(Program::getProgramDefinition)
+            .filter(pdef -> pdef.displayMode().equals(DisplayMode.PUBLIC))
+            .collect(ImmutableList.toImmutableList());
+
+    return applicationsFuture.thenApplyAsync(
+        applications -> {
+          checkForDuplicateDrafts(applications);
+          return relevantProgramsForApplicant(activePrograms, applications);
+        },
+        httpExecutionContext.current());
+  }
+
+  private RelevantPrograms relevantProgramsForApplicant(
+      ImmutableList<ProgramDefinition> activePrograms, ImmutableSet<Application> applications) {
+    ImmutableMap<String, ProgramDefinition> activeProgramNames =
+        ImmutableMap.copyOf(
+            activePrograms.stream()
+                .collect(Collectors.toMap(ProgramDefinition::adminName, pdef -> pdef)));
+
+    // When new versions of Programs are created, they have distinct IDs but retain the
+    // same adminName. In order to find the most recent draft / active application,
+    // we first group by the unique name rather than the ID.
+    Map<String, Map<LifecycleStage, Optional<Application>>> mostRecentApplicationsByProgram =
+        applications.stream()
+            .collect(
+                Collectors.groupingBy(
+                    a -> {
+                      return a.getProgram().getProgramDefinition().adminName();
+                    },
+                    Collectors.groupingBy(
+                        Application::getLifecycleStage,
+                        // In practice, we don't expect an applicant to have multiple applications
```

Comment:
This isn't the case for obsolete. maybe just note that.

Reply:
Reworded, thanks for noticing.